### PR TITLE
[Qwix] Quantize tgmm kernel 

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

Quantizing tgmm kernel used in backward pass of sparse matmul. 

e2e testing was done on DeepSeek-v3 with megablox=true and quantization=int8 on v5p-512. 

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Results are in this doc: http://shortn/_1OBSON0ioo

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/kernels/megablox/gmm.py [hunk 1]
```diff
@@ -720,23 +767,45 @@ def _do():
...
-          rhs_mask[...],
-          loaded_rhs.astype(jnp.float32),
-          jnp.zeros_like(rhs, jnp.float32),
-      )
-
-      acc_scratch[...] += lax.dot(
-          loaded_lhs.astype(input_dtype),
-          loaded_rhs.astype(input_dtype),
+      if use_qwix_quantization and isinstance(lhs, QArray):
+        loaded_lhs = load_qarray(lhs)
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py]: With latest Qwix, this can be written as `loaded_lhs = lhs[...]`.

File: MaxText/kernels/megablox/gmm.py [hunk 2]
```diff
@@ -815,6 +888,15 @@ def out_transform_indices(n_i, k_i, grid_id, group_metadata, group_offset):
       cost_estimate=cost_estimate,
   )
 
+  if use_qwix_quantization and lhs_quantize_dtype is not None:
+    lhs = qpl.quantize(lhs, qtype=lhs_quantize_dtype)
+    lhs = dataclasses.replace(lhs, scale=lhs.scale.astype(jnp.float32))
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py]: There's now a `scale_dtype` parameter to `qpl.quantize`, so we could write `lhs = qpl.quantize(lhs, qtype=lhs_quantize_dtype, scale_dtype=jnp.float32)`.

File: MaxText/layers/moe.py [hunk 3]
```diff
@@ -691,7 +691,11 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
         quant_dg = self.quant.quant_dg
         lhs_quantize_dtype = quant_dg.fwd.dg_quantizer.lhs.numerics.get_dtype()
         rhs_quantize_dtype = quant_dg.fwd.dg_quantizer.rhs.numerics.get_dtype()
-
+      if self.config.use_qwix_quantization:
+        quantization_rule = quantizations.get_quantization_rule(self.config)
```
Comment (liudangyi) [inline MaxText/layers/moe.py]: There's now a `qpl.get_current_rule` API that returns the current rule. The op name could just be `dot_general` or any other string if we want to configure it separately through `rule.op_names`.

File: MaxText/kernels/megablox/gmm.py [hunk 4]
```diff
@@ -825,7 +853,11 @@ def out_transform_indices(n_i, k_i, grid_id, group_metadata, group_offset):
   flops = 2 * m * k * n
   cost_estimate = pl.CostEstimate(flops=flops, bytes_accessed=bytes_accessed, transcendentals=0)
   lhs = lhs.swapaxes(0, 1)
-  call_gmm = pl.pallas_call(
+  if lhs_quantize_dtype is not None or rhs_quantize_dtype is not None:
```
Comment (RissyRan) [inline MaxText/kernels/megablox/gmm.py]: Do we also need ` use_qwix_quantization=True`?

Comment (khatwanimohit) [reply MaxText/kernels/megablox/gmm.py]: yes, thanks! updated

**Review by RissyRan — APPROVED** [review]
